lib/src/api_analysis/lower_bound_constraint_analysis.dart

```diff
@@ -0,0 +1,629 @@
+        cachePath: cachePath,
+        pubHostedUrl: pubHostedUrl,
+      ))
+          .last
+          .toString(),
+      destination: dummyPath,
+      pubHostedUrl: pubHostedUrl,
+      pubCachePath: pubCachePath,
+    );
+  } on ProcessException catch (exception) {
+    throw AnalysisException(
+        'Failed to download target package $targetName with error code ${exception.errorCode}: ${exception.message}');
+  }
+
+  // create session for analysing the package being searched for issues (the target package)
+  final collection = AnalysisContextCollection(includedPaths: [dummyPath]);
+  final context = PackageAnalysisContextWithStderr(
+    session: collection.contextFor(dummyPath).currentSession,
+    packagePath: dummyPath,
+    targetPackageName: targetName,
+  );
+
+  // if there are no dependencies, there are no issues
+  if (context.dependencies.isEmpty) {
+    return [];
+  }
+
+  final dependencySummaries = <String, PackageShape>{};
+
+  // iterate over each dependency of the target package and for each one:
+  // - determine minimum allowed version
+  // - determine installed (current/actual) version
+  // - download minimum allowed version
+  // - produce a summary of the minimum allowed version
+  for (final dependencyEntry in context.dependencies.entries) {
+    final dependencyName = dependencyEntry.key;
+    final dependencyVersionConstraint = dependencyEntry.value.version;
+    final dependencyDestination =
+        path.join(dependencyFolder, '${dependencyName}_dummy');
+
+    // determine the minimum allowed version of this dependency as allowed
+    // by the constraints imposed by the target package
+    final allVersions = await fetchSortedPackageVersionList(
+      packageName: dependencyName,
+      cachePath: cachePath,
+      pubHostedUrl: pubHostedUrl,
+    );
+    final minVersion =
+        allVersions.firstWhereOrNull(dependencyVersionConstraint.allows);
+
+    if (minVersion == null) {
+      context.warning(
+          'Skipping dependency $dependencyName, could not determine minimum allowed version.');
+      continue;
+    }
+
+    // download minimum allowed version of dependency
+    try {
+      await fetchUsingDummyPackage(
+        name: dependencyName,
+        version: minVersion.toString(),
+        destination: dependencyDestination,
+        pubHostedUrl: pubHostedUrl,
+        pubCachePath: pubCachePath,
+      );
+    } on ProcessException catch (exception) {
+      // it is expected that sometimes the SDK constraint on the lowest version of the dependency will be too low
+      if (!(exception.errorCode == 1 &&
+          exception.message.contains('which requires SDK version') &&
+          exception.message.endsWith('version solving failed.\n'))) {
+        context.warning(
+            'Skipping dependency $dependencyName of target package $targetName, failed to download it with error code ${exception.errorCode}: ${exception.message}');
+      }
+      continue;
```

Comment:
Shouldn't we just try with a higher version? How frequently does this happen?

Just because the lowest possible version doesn't work, doesn't mean that there isn't a problem. It's possible that a more recent version can be resolved, but still is missing features available in the latest version..

Reply:
> It's possible that a more recent version can be resolved, but still is missing features available in the latest version..
True, this is worth trying.

Reply:
started checking all allowed versions (except the installed one) to try to find one that passes version solving, beginning with the smallest